internal/btf/btf.go

```diff
@@ -193,67 +194,98 @@ func loadSpecFromELF(file *internal.SafeELFFile) (*Spec, error) {
 	oli := make(map[string]LineInfos)
 	ocr := make(map[string]CoreRelos)
 
 	for secName, secFuncs := range info.funcInfos {
 		// Collect functions from each section and organize them by name.
+		var funcs []*Func
 		for _, fi := range secFuncs {
-			name, err := fi.Name(spec)
+			typ, err := spec.TypeByID(fi.TypeID)
 			if err != nil {
-				return fmt.Errorf("looking up function name: %w", err)
+				return err
+			}
+
+			fn, ok := typ.(*Func)
+			if !ok {
+				return fmt.Errorf("type ID %d is a %T, but expected a Func", fi.TypeID, typ)
 			}
 
-			// FuncInfo offsets are scoped to the ELF section. Zero them out
-			// since they are meaningless outside of that context. The linker
-			// will determine the offset of the function within the final
-			// instruction stream before handing it off to the kernel.
-			fi.InsnOff = 0
+			// C doesn't have anonymous functions, but check just in case.
+			if fn.Name == "" {
+				return fmt.Errorf("func with type ID %d doesn't have a name", fi.TypeID)
+			}
+
+			ofi[fn.Name] = FuncInfo{fn}
+			funcs = append(funcs, fn)
+		}
 
-			ofi[name] = fi
+		// Consider an ELF section that contains 3 functions (a, b, c)
+		// at offsets 0, 10 and 15 respectively. Offset 5 will return function a,
+		// offset 12 will return b, offset >= 15 will return c, etc.
+		sort.Slice(secFuncs, func(i, j int) bool {
+			return secFuncs[i].InsnOff < secFuncs[j].InsnOff
+		})
+		funcForOffset := func(offset uint32) (fn *Func, fnOffset uint32) {
+			for i, fi := range secFuncs {
+				if fi.InsnOff > offset {
+					break
+				}
+				fn = funcs[i]
+				fnOffset = fi.InsnOff
+			}
+			return
 		}
 
 		// Attribute LineInfo records to their respective functions, if any.
 		if lines := info.lineInfos[secName]; lines != nil {
 			for _, li := range lines {
-				fi := secFuncs.funcForOffset(li.InsnOff)
-				if fi == nil {
+				fn, fnOffset := funcForOffset(li.InsnOff)
+				if fn == nil {
 					return fmt.Errorf("section %s: error looking up FuncInfo for LineInfo %v", secName, li)
 				}
 
-				// Offsets are ELF section-scoped, make them function-scoped by
-				// subtracting the function's start offset.
-				li.InsnOff -= fi.InsnOff
+				line, err := spec.strings.Lookup(li.LineOff)
+				if err != nil {
+					return fmt.Errorf("lookup of line: %w", err)
+				}
 
-				name, err := fi.Name(spec)
+				fileName, err := spec.strings.Lookup(li.FileNameOff)
 				if err != nil {
-					return fmt.Errorf("looking up function name: %w", err)
+					return fmt.Errorf("lookup of filename: %w", err)
 				}
 
-				oli[name] = append(oli[name], li)
+				lineNumber := li.LineCol >> 10
+				lineColumn := li.LineCol & 0x3ff
+
+				oli[fn.Name] = append(oli[fn.Name], LineInfo{
+					fileName,
+					line,
+					lineNumber,
+					lineColumn,
+					// Offsets are ELF section-scoped, make them function-scoped by
+					// subtracting the function's start offset.
+					li.InsnOff - fnOffset,
+					li.FileNameOff,
+					li.LineOff,
+				})
```

Comment:
True, the downside of using named keys is that you don't get a compile error when extending a struct definition. Since extension happens more often than reordering I tend to prefer this style for internal types.